### PR TITLE
relaunch `RangeBounds`

### PR DESCRIPTION
**Pull request by liigo**
based on PR #43033 (by @clarcharr ) which is out of time for a bit long time

r? @sfackler or @alexcrichton

### CODE REVIEW

**Comment by sfackler** [comment]
Thanks for updating this!

@bors r+

File: src/test/run-pass/sync-send-iterators-in-libcollections.rs [hunk 1]
```diff
@@ -50,11 +50,11 @@ fn main() {
     all_sync_send!(BinaryHeap::<usize>::new(), iter, drain, into_iter);
 
     all_sync_send!(BTreeMap::<usize, usize>::new(), iter, iter_mut, into_iter, keys, values);
-    is_sync_send!(BTreeMap::<usize, usize>::new(), range((Included(&0), Included(&9))));
-    is_sync_send!(BTreeMap::<usize, usize>::new(), range_mut((Included(&0), Included(&9))));
+    is_sync_send!(BTreeMap::<usize, usize>::new(), range((Included(0), Included(9))));
+    is_sync_send!(BTreeMap::<usize, usize>::new(), range_mut((Included(0), Included(9))));
```
Comment (alexcrichton) [inline src/test/run-pass/sync-send-iterators-in-libcollections.rs]: Was this change required to get this compiling? If so this may be a breaking change?

File: src/liballoc/btree/map.rs [hunk 2]
```diff
@@ -783,18 +782,18 @@ impl<K: Ord, V> BTreeMap<K, V> {
...
+    /// for (&key, &value) in map.range((Included(4), Included(8))) {
     ///     println!("{}: {}", key, value);
     /// }
     /// assert_eq!(Some((&5, &"b")), map.range(4..).next());
     /// ```
     #[stable(feature = "btree_range", since = "1.17.0")]
-    pub fn range<T: ?Sized, R>(&self, range: R) -> Range<K, V>
-        where T: Ord, K: Borrow<T>, R: RangeArgument<T>
+    pub fn range<T, R>(&self, range: R) -> Range<K, V>
+        where T: Ord, K: Borrow<T>, R: Into<RangeBounds<T>>
```
Comment (sfackler) [inline src/liballoc/btree/map.rs]: Shouldn't this be `RangeBounds<&T>`?

Comment (clarcharr) [reply src/liballoc/btree/map.rs]: So I actually recall this being one of the things I struggled with making the original PR. This prevents things like `"a".to_string()..="z".to_string()`. Because `Borrow` works differently from `AsRef`, this prevents both `String` ranges and `str` ranges from working properly.

There may be a way to make it work but I was really stumped in the original PR and got busy and didn't pick it back up.

Comment (liigo) [reply src/liballoc/btree/map.rs]: I'm stumped too.

**Comment by alexcrichton** [comment]
ping @liigo, just want to make sure this doesn't fall off your radar!

**Comment by alexcrichton** [comment]
Ok I'm going to close this due to inactivity, but feel free to resubmit @liigo!

**Comment by GuillaumeGomez** [comment]
I'd really like to get this stabilized so if @liigo isn't available to di it, I'll gladly do it.
